src/common/tools/time.cpp

```diff
@@ -4,10 +4,149 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#if defined(__ZEPHYR__)
+#include <zephyr/sys/timeutil.h>
+#endif
+
 #include "aos/common/tools/time.hpp"
 
 namespace aos {
 
+namespace {
+
+/***********************************************************************************************************************
+ * Constants
+ **********************************************************************************************************************/
+
+constexpr auto cUTCFormat = "%Y-%m-%dT%H:%M:%SZ";
+
+/***********************************************************************************************************************
+ * Static
+ **********************************************************************************************************************/
+
+const char* ConsumeChars(const char* s, char* dest, size_t cnt)
+{
+    if (strnlen(s, cnt) < cnt) {
+        return nullptr;
+    }
+
+    memcpy(dest, s, cnt);
+    dest[cnt] = '\0';
+
+    return s + cnt;
+}
+
+const char* ConsumeChar(const char* s, char ch)
+{
+    if (*s != ch) {
+        return nullptr;
+    }
+
+    return ++s;
+}
+
+const char* ConsumeDate(const char* s, struct tm* tm_time)
+{
+    char year[4 + 1];
+    char month[2 + 1];
+    char day[2 + 1];
+
+    s = ConsumeChars(s, year, 4);
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChar(s, '-');
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChars(s, month, 2);
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChar(s, '-');
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChars(s, day, 2);
+    if (!s) {
+        return nullptr;
+    }
+
+    tm_time->tm_year = atoi(year) - 1900;
+    tm_time->tm_mon  = atoi(month) - 1;
+    tm_time->tm_mday = atoi(day);
+
+    return s;
+}
+
+const char* ConsumeTime(const char* s, struct tm* tm_time)
+{
+    char hour[2 + 1];
+    char minute[2 + 1];
+    char second[2 + 1];
+
+    s = ConsumeChars(s, hour, 2);
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChar(s, ':');
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChars(s, minute, 2);
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChar(s, ':');
+    if (!s) {
+        return nullptr;
+    }
+
+    s = ConsumeChars(s, second, 2);
+    if (!s) {
+        return nullptr;
+    }
+
+    tm_time->tm_hour = atoi(hour);
+    tm_time->tm_min  = atoi(minute);
+    tm_time->tm_sec  = atoi(second);
+
+    return s;
+}
+
+Error ParseDateTime(const char* s, const char* format, struct tm* tm_time)
+{
+    if (strcmp(format, cUTCFormat) != 0) {
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    s = ConsumeDate(s, tm_time);
+    if (!s) {
+        return ErrorEnum::eFailed;
+    }
+
+    s = ConsumeChar(s, 'T');
+    if (!s) {
+        return ErrorEnum::eFailed;
+    }
+
+    s = ConsumeTime(s, tm_time);
+    if (!s) {
+        return ErrorEnum::eFailed;
+    }
```

Comment:
maybe will be helpfull to provide err message, not just fail to better understand the err reason?